### PR TITLE
Add recursive deletes for BlobContainers and fix Snapshot Deletion

### PR DESCRIPTION
**Pull request by mkleen**
This improves the snapshot deletion process making sure all files are removed when a snapshot is deleted. The problem was discovered while porting over https://github.com/elastic/elasticsearch/pull/43281.

It is based on the following steps:

- Add recursive deletes based on https://github.com/elastic/elasticsearch/pull/43281 implementing `blobContainer.delete()` to simplify the deletion process.
- Adding assertions for the Snapshot tests to make sure all files are removed.
- Adapt the relevant parts of the `BlobStoreRepository` according to the state of https://github.com/elastic/elasticsearch/pull/46250/.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -927,24 +930,7 @@ long readSnapshotIndexLatestBlob() throws IOException {
...
-                final long curr = Long.parseLong(blobName.substring(INDEX_FILE_PREFIX.length()));
-                latest = Math.max(latest, curr);
-            } catch (NumberFormatException nfe) {
-                // the index- blob wasn't of the format index-N where N is a number,
-                // no idea what this blob is but it doesn't belong in the repository!
-                LOGGER.debug("[{}] Unknown blob in the repository: {}", metadata.name(), blobName);
-            }
-        }
-        return latest;
+        return latestGeneration(blobContainer().listBlobsByPrefix(INDEX_FILE_PREFIX).keySet());
```
Comment (mkleen) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:933]: This was the part which had to be adapted.

File: server/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java [hunk 2]
```diff
@@ -555,11 +556,11 @@ private void assertSnapShotState(String snapShotName, SnapshotState state) {
         assertThat(response.rows()[0][1], is(1));
     }
 
-    private static void assertAllRepoSnapshotFilesAreDeleted(File location) {
+    private static void assertAllRepoSnapshotFilesAreDeleted(File location) throws IOException {
         //Make sure the file location does not consist of any .dat file
-        for (var file : location.listFiles()) {
-            assertThat(file.getName().endsWith(".dat"), is(false));
-        }
```
Comment (mkleen) [inline server/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java:562]: Make sure to check recursively.

File: blackbox/test_s3.py [hunk 3]
```diff
@@ -103,6 +107,11 @@ def tearDownClass(cls):
         crate_node.stop()
 
     def test_copy_to_s3_copy_from_s3_roundtrip(self):
+        client = Minio('127.0.0.1:9000',
```
Comment (mfussenegger) [inline blackbox/test_s3.py:110]: Should the client be closed eventually, or is that not necessary?

Comment (mkleen) [reply blackbox/test_s3.py:110]: It is not necessary according to https://github.com/minio/minio-py

**Comment by mkleen** [comment]
> Do we need a release note entry for this?

i thought about the same, I am not sure. What do you think ? Technically we fixed a bug.

**Comment by mfussenegger** [comment]
> > Do we need a release note entry for this?
> 
> i thought about the same, I am not sure. What do you think ? Technically we fixed a bug.

I'd suggest to add one.

File: docs/appendices/release-notes/unreleased.rst [hunk 4]
```diff
@@ -268,5 +268,8 @@ Performance improvements
 Fixes
 =====
 
+- Fixed an issue where :ref:`drop snapshots <ref-drop-snapshot>` would not
+  delete all the related data.
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst:272]: Maybe clarify that this affected Azure repositories? Or was it also the case for S3?

**Pull request merged**